### PR TITLE
Add Documentation for MulDepthAnalysis

### PR DESCRIPTION
**Pull request by Maokami**
This PR contains adding more comments in MulDepthAnalysis. (#780)

### CODE REVIEW

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h [hunk 1]
```diff
@@ -9,31 +9,75 @@
...
+///     \[
+///     \text{MulDepth}(\text{op}) =
+///     \begin{cases}
+///     \max\limits_{i}
+///     \left(\text{MulDepth}(\text{operand}_i(\text{op}))\right) + 1 & \text{if
+///     op is multiplication} \\ \max\limits_{i}
+///     \left(\text{MulDepth}(\text{operand}_i(\text{op}))\right) &
+///     \text{otherwise} \end{cases}
+///     \]
+
```
Comment (Maokami) [inline lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:46]: I'm not sure this latex comment can be supported?

Comment (j2kun) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:46]: It is not supported in the comments here, but if you put it in the tablegen, it will be rendered as latex on heir.dev

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h [hunk 2]
```diff
@@ -9,31 +9,75 @@
 namespace mlir {
 namespace heir {
 
+/// This class represents the multiplicative depth in an FHE program.
+///
+/// The multiplicative depth in an FHE program refers to the maximum number of
```
Comment (j2kun) [inline lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:14]: I would rephrase this more precisely:

> The multiplicative depth of a circuit refers to the maximum number
> of multiplication gates on any path from a circuit input to
> an output. In a more general FHE program, the multiplicative depth
> is the maximum number of ciphertext-ciphertext or plaintext-ciphertext
> multiplication operations on any use-def chain.
> 
> In FHE, multiplicative depth bounds the number of operations
> that can be sequentially performed on a ciphertext before the
> noise introduced during those operations renders the ciphertext 
> undecryptable.

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h (hunk 1 above)
Comment (j2kun) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:46]: What about this pseudocode instead?

```
MulDepth(z) = case
  mul_op(x, y):
    max(MulDepth(x), MulDepth(y)) + 1
  any_op(operands):
    max(map(MulDepth, operands))
```

**Review by j2kun — APPROVED** [review]
Minor suggestions only!

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h (hunk 1 above)
Comment (Maokami) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:46]: nice!

**Review by j2kun — APPROVED** [review]

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h [hunk 3]
```diff
@@ -9,31 +9,68 @@
...
   bool operator==(const MulDepth &rhs) const { return value == rhs.value; }
 
+  /// Join two multiplicative depth.
   static MulDepth join(const MulDepth &lhs, const MulDepth &rhs) {
     if (!lhs.isInitialized()) return rhs;
     if (!rhs.isInitialized()) return lhs;
+    // If both are initialized, the larger value should be chosen for a sound
+    // analysis of the multiplicative depth. If both are initialized, the larger
+    // value between the two should be selected to soundly analyze the
+    // multiplicative depth.
```
Comment (asraa) [inline lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h:65]: Maybe this was a copy-paste thing?

```suggestion
    // If both are initialized, the larger value should be chosen for a sound
    // analysis of the multiplicative depth.
```

**Review by asraa — COMMENTED** [review]
Looks great!

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h (hunk 3 above)
Comment (Maokami) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h]: Oops!

Comment (Maokami) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h]: @asraa Could you possibly squash merge this commit? Otherwise, I can do it after 12 hours :)

**Comment by AlexanderViand-Intel** [comment]
Thanks for putting this together!

File: lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h (hunk 3 above)
Comment (asraa) [reply lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h]: done!
